zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+
+    /// Return the maximum allowed depth of the tree.
+    fn max_depth() -> u32;
+
+    /// Return the maximum allowed amount of nodes in the tree.
+    fn max_node_count() -> u32;
+}
+
+impl<T, U> LiquidityTreeHelper<T> for LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    fn propagate_fees_to_node(&mut self, index: u32) -> DispatchResult {
+        let path = self.path_to_node(index)?;
+        for i in path {
+            self.propagate_fees(i)?;
+        }
+        Ok(())
+    }
+
+    fn propagate_fees(&mut self, index: u32) -> DispatchResult {
+        let node = self.get_node(index)?;
+        if node.total_stake()? == Zero::zero() {
+            return Ok(()); // Don't propagate if there are no LPs under this node.
+        }
+        // Temporary storage to ensure that the borrow checker doesn't get upset.
+        let descendant_stake = node.descendant_stake;
+        if node.is_leaf() {
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&node.lazy_fees)?;
+                Ok(())
+            })?;
+        } else {
+            // The lazy fees that will be propagated down the tree.
+            let mut remaining_lazy_fees =
+                node.descendant_stake.bmul_bdiv(node.lazy_fees, node.total_stake()?)?;
+            // The fees that stay at this node.
+            let fees = node.lazy_fees.checked_sub_res(&remaining_lazy_fees)?;
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&fees)?;
+                Ok(())
+            })?;
+            let child_indices = self.children(index)?;
+            // Unwrap below can't fail since we're unwrapping from an array of length 2.
+            match child_indices.get(0).unwrap_or(&None) {
+                Some(lhs_index) => {
+                    self.mutate_node(*lhs_index, |lhs_node| {
+                        let child_lazy_fees = lhs_node
+                            .total_stake()?
+                            .bmul_bdiv(remaining_lazy_fees, descendant_stake)?;
+                        lhs_node.lazy_fees =
+                            lhs_node.lazy_fees.checked_add_res(&child_lazy_fees)?;
+                        remaining_lazy_fees =
+                            remaining_lazy_fees.checked_sub_res(&child_lazy_fees)?;
+                        Ok(())
+                    })?;
+                }
+                None => (),
+            }
+            // Unwrap below can't fail since we're unwrapping from an array of length 2.
+            match child_indices.get(1).unwrap_or(&None) {
+                Some(rhs_index) => {
+                    self.mutate_node(*rhs_index, |rhs_node| {
+                        rhs_node.lazy_fees =
+                            rhs_node.lazy_fees.checked_add_res(&remaining_lazy_fees)?;
+                        Ok(())
+                    })?;
+                }
+                None => (),
+            }
```

Comment:
It's just that for the left-hand side child you know exactly the fees to charge, but for the right-hand side node you just add the remainder. So my suggestion went into the direction to calculate the fees to charge for the node on the right-hand side too.

Reply:
> [...] but for the right-hand side node you just add the remainder.

Yeah, the right child is the last of the three nodes in question to receive fees, so it should receive the rest. Are you concerned this might be wrong?

> So my suggestion went into the direction to calculate the fees to charge for the node on the right-hand side too.

There's a problem with that. If you use a ratio to calculate both the portion of the left node's fees and the right node's fees, then you might end up rounding up twice, which will result in you needing to distribute one Pennock more than you have.

Generally, when using fixed point operations, addition and subtraction is preferred because they are always 100% precise. If you use multiplication to split up a balance, always do multiply-subtract-multiply-substract. You can see us take this approach in many situations (always when we're calculating fees), like in the neo-swaps pallet.

Reply:
> Correct me if I am wrong, but the node 1 gets 20 total stake of node 1 divided by 23 descendant stake of node 0 (from the rust code above) of the remaining fees and the node 2 gets just the rest here (3 / 23) (the node 2 has a total stake of 3).

Sorry, missed that. Yeah, the left node gets 20/23 of the lazy_fees that are being propagated down and the right node received the remaining 3/23.

Reply:
> Are you concerned this might be wrong?

Yep, would be better to verify we have the correct remainder.

Reply:
The proof is kind of as follows. 

The left child receives `a = lazy_fees * left_child.stake / node.descendant_stake` and the right child receives `b = lazy_fees * right_child.stake / node.descendant_stake`. Note that `descendant_stake = left_child.stake + right_child.stake`. Thus,

```
lazy_fees = (left_child.stake + right_child.stake) / (left_child.stake + right_child.stake) = (left_child.stake + right_child.stake) / node.descendant_stake = left_child.stake / node.descendant_stake + right_child.stake / node.descendant_stake = a + b.
```

Thus, `remainder = lazy_fees - a = b`. Thus, the right node receiving the remainder is correct.